Test/Added test of components and implemented snapshot tests

matiasbavera opened the pull request:
Fixes #49

## What's new
* Implemented snapshot tests: Snapshot tests are a very useful tool whenever you want to make sure your UI does not change unexpectedly.
* Improved ~4% test coverage
* Added tests for every type of Doors
* Added tests for doors lifts places and robots overlays.
* Added tests for utils functions.


A typical snapshot test case for a web app renders a UI component, takes a snapshot, then compares it to a reference snapshot file stored alongside the test. The test will fail if the two snapshots do not match: either the change is unexpected, or the reference snapshot needs to be updated to the new version of the UI component.

baviera08 commented:
```
-----------------------------------------|----------|----------|----------|----------|-------------------|
File                                     |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
-----------------------------------------|----------|----------|----------|----------|-------------------|
All files                                |    52.63 |    46.95 |    50.85 |    52.44 |                   |
 src                                     |    26.09 |    18.07 |    32.81 |    26.21 |                   |
  app-config.ts                          |    26.67 |       25 |       50 |    26.67 |... 26,27,30,31,32 |
  auth-service.ts                        |        0 |        0 |        0 |        0 |... ,93,98,101,104 |
  dispenser-state-manager.ts             |       40 |      100 |    33.33 |       40 |          10,15,16 |
  door-state-manager.ts                  |       60 |      100 |    66.67 |       60 |             15,16 |
  fleet-manager.ts                       |       60 |      100 |    66.67 |       60 |             15,16 |
  index.tsx                              |        0 |      100 |      100 |        0 |           9,11,19 |
  lift-state-manager.ts                  |    72.09 |    52.38 |    91.67 |    72.09 |... ,54,60,109,120 |
  robot-trajectory-manager.ts            |       15 |        0 |     7.14 |       15 |... 34,138,139,140 |
  serviceWorker.ts                       |        0 |        0 |        0 |        0 |... 31,138,139,140 |
  settings.ts                            |    71.43 |       50 |    66.67 |    71.43 |             22,30 |
 src/components                          |    64.88 |     55.7 |    64.29 |     65.4 |                   |
  app.tsx                                |    56.16 |    72.73 |     45.1 |    55.73 |... 32,405,406,408 |
  disableable-typography.tsx             |      100 |      100 |      100 |      100 |                   |
  dispenser-item.tsx                     |    90.91 |       75 |      100 |    90.48 |             39,64 |
  dispensers-panel.tsx                   |    57.89 |       25 |    57.14 |    64.71 | 20,21,22,24,28,40 |
  door-item.tsx                          |    55.88 |       40 |      100 |    54.55 |... 63,165,167,178 |
  doors-panel.tsx                        |       68 |     62.5 |       75 |    73.91 | 36,37,38,40,44,61 |
  loading-screen.tsx                     |      100 |    57.14 |      100 |      100 |          28,31,34 |
  login.tsx                              |        0 |      100 |        0 |        0 |... 68,71,77,78,85 |
  main-menu.tsx                          |      100 |      100 |      100 |      100 |                   |
  omni-panel-view.tsx                    |      100 |      100 |      100 |      100 |                   |
  omni-panel.tsx                         |    81.82 |    16.67 |    66.67 |       80 |             50,54 |
  place-item.tsx                         |      100 |      100 |      100 |      100 |                   |
  places-panel.tsx                       |    57.89 |       50 |    66.67 |    64.71 | 21,22,23,25,29,44 |
  robot-item-form.tsx                    |    80.39 |    65.79 |    69.23 |       80 |... ,52,87,107,124 |
  robot-item-information.tsx             |      100 |      100 |      100 |      100 |                   |
  robot-item.tsx                         |       65 |    45.45 |       75 |    63.16 |... 14,116,120,124 |
  robots-panel.tsx                       |    66.67 |       50 |       75 |    68.18 |... 67,68,69,71,75 |
  settings-drawer.tsx                    |    83.33 |       25 |       80 |    81.82 |             34,35 |
  spotlight-value.tsx                    |        0 |        0 |        0 |        0 |                   |
  tab.tsx                                |      100 |      100 |      100 |      100 |                   |
 src/components/lift-item                |    76.85 |     56.6 |     72.5 |       77 |                   |
  lift-item-form.tsx                     |    73.47 |       25 |    70.59 |    71.74 |... 4,73,75,93,111 |
  lift-item-information.tsx              |      100 |      100 |      100 |      100 |                   |
  lift-item.tsx                          |       90 |    81.82 |    71.43 |    88.89 |             54,58 |
  lifts-panel.tsx                        |    56.52 |       40 |       60 |     61.9 |... 46,47,49,53,72 |
 src/components/schedule-visualizer      |    36.91 |    47.52 |    32.04 |    36.21 |                   |
  arrow.tsx                              |      100 |      100 |      100 |      100 |                   |
  colors.ts                              |     3.85 |        0 |        0 |     3.85 |... 34,36,40,41,47 |
  doors-overlay.tsx                      |       90 |     62.5 |       75 |       90 |                34 |
  index.tsx                              |        0 |        0 |        0 |        0 |... 35,238,255,257 |
  lift-overlay.tsx                       |    85.71 |       50 |    66.67 |    85.71 |                26 |
  lift.tsx                               |    98.31 |    93.48 |      100 |      100 |          40,58,59 |
  place.tsx                              |      100 |      100 |      100 |      100 |                   |
  places-overlay.tsx                     |    83.33 |        0 |    66.67 |    83.33 |                25 |
  robot-trajectories-overlay.tsx         |        0 |        0 |        0 |        0 |... 24,27,29,33,49 |
  robot-trajectory.tsx                   |    58.33 |       50 |    57.14 |    60.87 |... 71,73,74,76,82 |
  robot.tsx                              |       80 |     62.5 |    71.43 |    78.57 |          41,42,64 |
  robots-overlay.tsx                     |    77.78 |       25 |    33.33 |    77.78 |             27,33 |
  svg-overlay.tsx                        |      100 |      100 |      100 |      100 |                   |
  trajectory-animations.tsx              |        0 |        0 |        0 |        0 |... 03,214,224,236 |
 src/components/schedule-visualizer/door |    98.08 |    76.19 |       90 |    98.04 |                   |
  door-default.tsx                       |      100 |      100 |      100 |      100 |                   |
  door-double-hinge.tsx                  |      100 |      100 |      100 |      100 |                   |
  door-double-slide.tsx                  |      100 |      100 |      100 |      100 |                   |
  door-single-hinge.tsx                  |      100 |      100 |      100 |      100 |                   |
  door-single-slide.tsx                  |      100 |      100 |      100 |      100 |                   |
  door-style.tsx                         |      100 |      100 |      100 |      100 |                   |
  door.tsx                               |    93.33 |    70.59 |    66.67 |    93.33 |                60 |
 src/leaflet                             |        0 |      100 |        0 |        0 |                   |
  control-positions.ts                   |        0 |      100 |        0 |        0 |... 15,20,25,30,35 |
 src/mock                                |    51.11 |       75 |    17.65 |    51.11 |                   |
  fake-auth-service.ts                   |        0 |      100 |        0 |        0 |        5,13,17,21 |
  fake-traj-manager.ts                   |    71.43 |       50 |       50 |    71.43 |             18,26 |
  fake-transport.ts                      |    52.94 |       80 |    18.18 |    52.94 |... 59,70,71,88,94 |
 src/mock/data                           |    84.62 |       50 |      100 |       84 |                   |
  building-map-airport.ts                |      100 |      100 |      100 |      100 |                   |
  building-map-office.ts                 |      100 |      100 |      100 |      100 |                   |
  building-map.ts                        |    55.56 |       50 |      100 |    55.56 |       15,16,17,18 |
  dispenser-states.ts                    |      100 |      100 |      100 |      100 |                   |
  door-states.ts                         |      100 |      100 |      100 |      100 |                   |
  fleets.ts                              |      100 |      100 |      100 |      100 |                   |
  lift-states.ts                         |      100 |      100 |      100 |      100 |                   |
  places.ts                              |      100 |      100 |      100 |      100 |                   |
 src/util                                |    69.81 |    14.29 |    57.14 |    68.63 |                   |
  alerts.ts                              |       50 |        0 |        0 |       50 |              4,13 |
  angle-calculation.ts                   |      100 |      100 |      100 |      100 |                   |
  css-utils.ts                           |      100 |      100 |      100 |      100 |                   |
  cublic-spline.ts                       |    70.73 |        0 |    66.67 |    69.23 |... 36,137,138,140 |
  index.ts                               |        0 |      100 |        0 |        0 |               2,3 |
-----------------------------------------|----------|----------|----------|----------|-------------------|

```

baviera08 commented:
I can't assign a reviewer with GH, so 
reviewer: @koonpeng

koonpeng commented on src/components/schedule-visualizer/tests/places-overlay.test.tsx:
```diff
@@ -0,0 +1,25 @@
+import L from 'leaflet';
+import { Map as LMap } from 'react-leaflet';
+import PlacesOverlay from '../places-overlay';
+import Place from '../place';
```
Places is not used in RMF anymore and is removed from in a recent PR, you can remove this test as well.

matiasbavera replied:
Done, thanks!

koonpeng commented on src/components/tests/dispenser-item.test.tsx:
```diff
@@ -4,7 +4,7 @@ import * as RomiCore from '@osrf/romi-js-core-interfaces';
 import fakeDispenserStates from '../../mock/data/dispenser-states';
 import DispenserItem from '../dispenser-item';
-import DispenserPanel from '../dispensers-panel';
+import renderer from 'react-test-renderer';
```
Can you explain why we need to use `react-test-renderer` here and not in the other tests?

koonpeng commented on src/util/tests/angle-calculation.test.ts:
```diff
@@ -0,0 +1,8 @@
+import { radiansToDegrees } from "../angle-calculation"
+
+test('Convert correctly radians to degrees', () => {
```
Would floating point deviations cause this test to fail? Should we specify the number of decimals that we should check for?

matiasbavera replied:
Yeah, better safe than sorry. I added `toBeCloseTo`  to compare floating-point https://jestjs.io/docs/en/using-matchers#numbers

matiasbavera commented on src/components/schedule-visualizer/door/tests/door.test.tsx:
```diff
@@ -17,7 +23,68 @@ const mainDoor = {
+  Component: (props: DoorProps) => React.ReactElement,
+  door: RomiCore.Door,
+  currentMode: number,
+): JSX.Element => {
```
Yeah, makes sense. Interesting link https://stackoverflow.com/questions/58123398/when-to-use-jsx-element-vs-reactnode-vs-reactelement

matiasbavera replied:
Thanks! Done!

matiasbavera commented on src/components/schedule-visualizer/door/tests/door.test.tsx:
```diff
@@ -1,10 +1,16 @@
 import * as RomiCore from '@osrf/romi-js-core-interfaces';
 import { createMount } from '@material-ui/core/test-utils';
-import { ReactWrapper } from 'enzyme';
+import { ReactWrapper, shallow } from 'enzyme';
```
You're right it's a leftover. 
Shallow renders the root node and returns a shallow wrapper around it.
For more info: https://enzymejs.github.io/enzyme/docs/api/ShallowWrapper/shallow.html

matiasbavera commented on src/components/schedule-visualizer/door/tests/door.test.tsx:
```diff
@@ -43,60 +110,68 @@ test('Trigger click event', async () => {
+    doorClose: 'doorCloseStyle',
+    doorProcess: 'doorProcessStyle',
   };
+  const open = getDoorStyle(doorStyle, RomiCore.DoorMode.MODE_OPEN);
```
Good to know. Thanks! Changed!

matiasbavera commented on src/components/schedule-visualizer/door/tests/door.test.tsx:
```diff
@@ -43,60 +110,68 @@ test('Trigger click event', async () => {
-    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
-    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorClose)-(\d+)/)).toBe(true);
-    wrapper.unmount();
+  test('Style of DobleHingeDoors', () => {
```
You're right! Thank you!

matiasbavera commented on src/components/schedule-visualizer/tests/doors-overlay.test.tsx:
```diff
@@ -0,0 +1,51 @@
+import DoubleSlideDoor from '../door/door-double-slide';
+import SingleSlideDoor from '../door/door-single-slide';
+
+// 1 DOOR_TYPE_SINGLE_SLIDING
+// 1 DOOR_TYPE_DOUBLE_SLIDING
```
No, but I removed them and did something more understandable
`door_type: RomiCore.Door.DOOR_TYPE_DOUBLE_SLIDING,`
and 
`door_type: RomiCore.Door.DOOR_TYPE_SINGLE_SLIDING,`

matiasbavera commented on src/components/schedule-visualizer/tests/place.test.tsx:
```diff
@@ -1,13 +1,14 @@
-import React from 'react';
-import getBuildingMap from '../../../mock/data/building-map';
+import officeMap from '../../../mock/data/building-map-office';
 import Place from '../place';
```
Removed!

matiasbavera commented on src/components/schedule-visualizer/tests/lift.test.tsx:
```diff
@@ -4,19 +4,28 @@ import fakeLiftStates from '../../../mock/data/lift-states';
   currentFloor: string = 'L1',
-) => {
-  const wrapper = mount(
+): JSX.Element => {
```
Done, thanks!

matiasbavera commented on src/components/schedule-visualizer/tests/doors-overlay.test.tsx:
```diff
@@ -0,0 +1,51 @@
+];
+
+test('Render doors correctly', () => {
+  let clicked = false;
```
Yeah, removed. Thanks!

matiasbavera commented:
> In the tests we are using both enzyme's mount and react-test-renderer's render, can you explain why we need both of them?
Yeah actually we don't, I removed `react-test-renderer` and now I'm generating a Json from the enzyme wrapper for snapshot tests. I added another library to transform the wrapper to JSON though.
